internal/invocation/invocation_service.go

```diff
@@ -17,68 +17,143 @@
 	// removeCh carries correlationIDs to be removed
-	removeCh    <-chan int64
+	removeCh    chan int64
 	doneCh      chan struct{}
 	invocations map[int64]Invocation
 	handler     Handler
 	logger      ilogger.Logger
+	state       int32
 }
 
 func NewService(
-	requestCh <-chan Invocation,
-	urgentRequestCh <-chan Invocation,
-	responseCh <-chan *proto.ClientMessage,
-	removeCh <-chan int64,
 	handler Handler,
 	logger ilogger.Logger) *Service {
 	service := &Service{
-		requestCh:       requestCh,
-		urgentRequestCh: urgentRequestCh,
-		responseCh:      responseCh,
-		removeCh:        removeCh,
+		requestCh:       make(chan Invocation),
+		urgentRequestCh: make(chan Invocation),
+		responseCh:      make(chan *proto.ClientMessage),
+		removeCh:        make(chan int64),
 		doneCh:          make(chan struct{}),
 		invocations:     map[int64]Invocation{},
 		handler:         handler,
 		logger:          logger,
+		state:           ready,
 	}
 	go service.processIncoming()
 	return service
 }
 
 func (s *Service) Stop() {
+	if !atomic.CompareAndSwapInt32(&s.state, ready, stopped) {
+		return
+	}
 	close(s.doneCh)
 }
 
 func (s *Service) SetHandler(handler Handler) {
 	s.handler = handler
 }
 
+func (s *Service) SendRequest(ctx context.Context, inv Invocation) error {
+	if atomic.LoadInt32(&s.state) != ready {
+		return cb.WrapNonRetryableError(hzerrors.ErrClientNotActive)
+	}
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("sending invocation: %w", ctx.Err())
+	case <-s.doneCh:
+		return cb.WrapNonRetryableError(fmt.Errorf("sending invocation: %w", hzerrors.ErrClientNotActive))
+	case s.requestCh <- inv:
+		return nil
+	}
+}
+
+func (s *Service) SendUrgentRequest(ctx context.Context, inv Invocation) error {
+	if atomic.LoadInt32(&s.state) != ready {
+		return cb.WrapNonRetryableError(hzerrors.ErrClientNotActive)
+	}
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("sending urgent invocation: %w", ctx.Err())
+	case <-s.doneCh:
+		return cb.WrapNonRetryableError(fmt.Errorf("sending urgent invocation: %w", hzerrors.ErrClientNotActive))
+	case s.urgentRequestCh <- inv:
+		return nil
+	}
+}
+
+func (s *Service) WriteResponse(msg *proto.ClientMessage) error {
+	if atomic.LoadInt32(&s.state) != ready {
+		return cb.WrapNonRetryableError(hzerrors.ErrClientNotActive)
+	}
+	select {
+	case <-s.doneCh:
+		return cb.WrapNonRetryableError(fmt.Errorf("writing response: %w", hzerrors.ErrClientNotActive))
+	case s.responseCh <- msg:
+		return nil
+	}
+}
+
+func (s *Service) Remove(correlationID int64) error {
+	if atomic.LoadInt32(&s.state) != ready {
+		return cb.WrapNonRetryableError(hzerrors.ErrClientNotActive)
+	}
+	select {
+	case <-s.doneCh:
+		return cb.WrapNonRetryableError(fmt.Errorf("removing correlation: %w", hzerrors.ErrClientNotActive))
+	case s.removeCh <- correlationID:
+		return nil
+	}
+}
+
 func (s *Service) processIncoming() {
 loop:
 	for {
 		select {
-		case inv := <-s.requestCh:
+		case inv, ok := <-s.requestCh:
+			if !ok {
```

Comment:
We don't need these checks anymore since we are not closing these channels.

Reply:
OK: https://github.com/hazelcast/hazelcast-go-client/pull/655/commits/348b8d64bf5b9b85f73f01a47bad2d244cf0ef98